### PR TITLE
Ballot w abstain

### PR DESCRIPTION
**Pull request by jimscarver**

### CODE REVIEW

**Comment by jimscarver** [comment]
I don't see how to make this a draft pull request. I made the pull request to see if netify could build it.
```
(zulip-py3-venv) jim@DESKTOP-O6G6UAP:~/rv2020$ npm run build

> rv2020@0.4.0 build /home/jim/rv2020
> snowpack build --polyfill-node

[snowpack] ! building source…
[snowpack] ✔ build complete [0.03s]
[snowpack] ! installing dependencies…
[snowpack] ✔ install complete [3.50s]
[snowpack] ! minifying javascript...

Error: Transform failed with 1 error:
<stdin>:187:12: error: Expected ";" but found "rho"
    at failureErrorWithLog (/home/jim/rv2020/node_modules/esbuild/lib/main.js:458:15)
    at /home/jim/rv2020/node_modules/esbuild/lib/main.js:411:31
    at handleIncomingPacket (/home/jim/rv2020/node_modules/esbuild/lib/main.js:372:9)
    at Socket.readFromStdout (/home/jim/rv2020/node_modules/esbuild/lib/main.js:320:7)
    at Socket.emit (events.js:311:20)
    at addChunk (_stream_readable.js:294:12)
    at readableAddChunk (_stream_readable.js:275:11)
    at Socket.Readable.push (_stream_readable.js:209:10)
    at Pipe.onStreamRead (internal/stream_base_commons.js:186:23)

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! rv2020@0.4.0 build: `snowpack build --polyfill-node`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the rv2020@0.4.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/jim/.npm/_logs/2020-09-29T23_23_12_848Z-debug.log
```

File: src/web/ballot.js [hunk 1]
```diff
@@ -179,6 +181,30 @@ async function registryLookup(uri, httpUrl, { rnodeHttp, setStatus }) {
     return rhoExprToJS(result);
 }
 
+/** @type {(uri: string) => string } */
+export function lookup_ballot_user_rho(acct, balloturi, votersuri) {
+    return `new return ,
+    lookup(`rho:registry:lookup`)
```
Comment (dckc) [inline src/web/ballot.js]: The back-ticks here need escaping.

File: public/index.html [hunk 2]
```diff
@@ -25,7 +25,7 @@
             <img alt="RChain Community" src="https://avatars1.githubusercontent.com/u/43198800?s=88&v=4">
           </a>
         </div>
-        <div class="progress" style="width: 50%">
+        <div class="progress" style="width: 20%">
```
Comment (dckc) [inline public/index.html]: Width again? Looks like you didn't start this branch from master after your earlier work was merged.

File: package.json [hunk 3]
```diff
@@ -28,6 +28,7 @@
     "realclean": "rm -rf dist build node_modules web_modules",
     "start": "snowpack dev --polyfill-node",
     "build": "snowpack build --polyfill-node",
+    "fix": "snowpack --reload --polyfill-node",
```
Comment (dckc) [inline package.json]: I don't quite understand what this is about. Care to elaborate?

Not critical. It looks reasonably harmless.

File: public/index.html [hunk 4]
```diff
@@ -40,13 +40,12 @@
     <div class="app container">
       <div class="testNet">
         <div>
-          <br>
-          <br>
```
Comment (dckc) [inline public/index.html]: Why delete the brs?

Maybe this is just a draft...

File: public/index.html [hunk 5]
```diff
@@ -40,13 +40,12 @@
...
-          <h2>RChain Annual Meeting 2020 Ballot <sup>(Beta Test2)</sup></h2>
+          <h4>RChain Annual Meeting 2020 Ballot <sup>(Beta Test2)</sup></h4>
           <hr />
           <button id='startTour' class="btn btn-primary">Show Me How</button>
+          <br>
         </div>
-        <ol><small><small>
```
Comment (dckc) [inline public/index.html]: Ditto

File: public/index.html [hunk 6]
```diff
@@ -40,13 +40,12 @@
     <div class="app container">
       <div class="testNet">
         <div>
-          <br>
-          <br>
-          <h2>RChain Annual Meeting 2020 Ballot <sup>(Beta Test2)</sup></h2>
+          <h4>RChain Annual Meeting 2020 Ballot <sup>(Beta Test2)</sup></h4>
```
Comment (dckc) [inline public/index.html]: It needs to be h1, not h4 or h2

File: src/cli/voters [hunk 7]
```diff
@@ -0,0 +1,111 @@
+111122ww54CKibkmuS9dGZm98uAAu48wtrgBLyH4bsTdGhWCwucEp
```
Comment (dckc) [inline src/cli/voters:1]: Hmm... Checking in the voter list... Usually derived stuff isn't checked in.

Perhaps I missed an explanation for this exception

File: src/cli/isreg.sh [hunk 8]
```diff
@@ -0,0 +1,42 @@
+#usage: ./explore.sh URI ['{code}']
```
Comment (dckc) [inline src/cli/isreg.sh:1]: Please find another reviewer for the bash code.

File: src/web/ballot.js [hunk 9]
```diff
@@ -98,7 +100,7 @@ export function buildUI({ ethereumAddress, getElementById, querySelectorAll, cre
         const misc = { name: 'testNet', http: null, httpsAdmin: null }; // typechecker says we need these; runtime says we don't
         const node = getNodeUrls({ ...misc, ...testNet.readOnlys[0] });
         const { rnodeHttp } = rnodeWeb;
-        registryLookup(ui.agendaURI.value, node.httpUrl, { rnodeHttp, setStatus }).then(qas => {
+        registryLookup(ui.agendaURI.value,  node.httpUrl, { rnodeHttp, setStatus }).then(qas => {
```
Comment (dckc) [inline src/web/ballot.js]: Extra space won't pass lunt check.

File: src/web/ballot.js [hunk 10]
```diff
@@ -226,47 +252,51 @@ function showQuestions(qas, questionList) {
           <tr><td>${id}</td>
           <td>${shortDesc}
            ${docLink ? html`<br />see: <a href=${docLink} target="_blank">${id}</a>` : ''}</td>
-          ${radio(noAddr)} ${radio('', {checked: 'checked'})} ${radio(yesAddr)}
+          ${radio(noAddr)} ${radio(abstainAddr, {checked: 'checked'})} ${radio(yesAddr)}
```
Comment (dckc) [inline src/web/ballot.js:256]: Looks good

File: src/web/ballot.js [hunk 11]
```diff
@@ -226,47 +252,51 @@ function showQuestions(qas, questionList) {
...
           </dd>`;
     });
     m.render(questionList, markup);
 }
 
 /**
- * @typedef {{[refID: string]: { shortDesc: string, docLink?: string, yesAddr: string, noAddr: string }}} QAs
+ * @typedef {{[refID: string]: { shortDesc: string, docLink?: string, yesAddr: string, noAddr: string, abstainAddr: string }}} QAs
```
Comment (dckc) [inline src/web/ballot.js:263]: Good

File: src/web/ballot.js [hunk 12]
```diff
@@ -11,6 +11,8 @@ import { getAddrFromEth } from '../vendor/rnode-client-js/src/rev-address';
 import { transferMulti_rho } from '../rho/transfer-multi';
 import { lookup_rho } from '../rho/lookup';
 
+const voteruri = "rho:id:1ri71weozwuoanef9zit5p7ooafkmkzhkwo6phgaourcknbmi6ke7t";
```
Comment (dckc) [inline src/web/ballot.js]: This isn't really const, is it? It will change before we're done. Please open an issue to address this in a later PR.

**Review by dckc — CHANGES_REQUESTED** [review]
Backticks are critical, as netlify pointed out.

**Comment by jimscarver** [comment]
Yippie! account registered status and abstain voting works! Mostly only cosmetic stuff is left.

File: src/web/ballot.js (hunk 12 above)
Comment (jimscarver) [reply src/web/ballot.js]: https://github.com/rchain-community/rv2020/issues/98

File: src/web/ballot.js (hunk 9 above)
Comment (jimscarver) [reply src/web/ballot.js]: fixed.

File: src/web/ballot.js [hunk 13]
```diff
@@ -191,10 +199,10 @@ export function lookup_ballot_user_rho(acct, balloturi, votersuri) {
       for (@ballot <- valueCh) {
           lookup!( \`${votersuri}\` , *valueCh) |
           for (@accts <- valueCh) {   
-            return!({"registered": accts.contain("${acct}") ,"ballot": ballot})
+            return!({"registered": accts.contains("${acct}") ,"ballot": ballot})
```
Comment (dckc) [inline src/web/ballot.js:202]: interpolating `acct` this way risks injection. Try this instead:

```suggestion
            return!({"registered": accts.contains(${JSON.stringify(acct)}) ,"ballot": ballot})
```

**Pull request merged**